### PR TITLE
Solving IsSymbol instances

### PR DESCRIPTION
**Pull request by LiamGoodacre**
Opening this PR for discussion: there are some TODOs and other things.

Currently many of the `example` tests fail since I updated `tests/support/bower.json` to specify the current latest versions of libraries.  They fail with `The foreign module implementation for module Unsafe.Coerce is missing.`.  I added `purescript-unsafe-coerce` as a dependency, but it didn't help.  So I'm not sure why that fails.

However my `SolvingIsSymbol` test does pass 👍

### CODE REVIEW

File: src/Language/PureScript/TypeClassDictionaries.hs [hunk 1]
```diff
@@ -8,10 +8,10 @@ import Language.PureScript.Types
 -- |
 -- Data representing a type class dictionary which is in scope
 --
-data TypeClassDictionaryInScope
+data TypeClassDictionaryInScope n
   = TypeClassDictionaryInScope {
     -- | The identifier with which the dictionary can be accessed at runtime
-      tcdName :: Qualified Ident
+      tcdName :: n
```
Comment (paf31) [inline src/Language/PureScript/TypeClassDictionaries.hs]: Could we just make this into an `Expr`?

**Comment by paf31** [comment]
Looks great, thanks!

Yes, you'll have some issues with library versions if you update them all to v2.0 like that. There is an issue to track the update that needs to happen (#2400) but it's probably quite a bit of work. Maybe you can avoid it for now?

File: src/Language/PureScript/TypeClassDictionaries.hs (hunk 1 above)
Comment (LiamGoodacre) [inline src/Language/PureScript/TypeClassDictionaries.hs]: I think I tried something like that previously (think it was `Either Expr (Qualified Ident)`) but it caused a module dependency cycle.  I'll have another look later today.

**Comment by LiamGoodacre** [comment]
In that case I should probably use `purescript-symbols` at version `1.0.1` and drop the `TypeConcat` example.

File: examples/passing/SolvingIsSymbol.purs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+module Main where
+
+import Prelude
+import Data.Symbol
+import Control.Monad.Eff
+import Control.Monad.Eff.Console
+
+main = do
+  log (reflectSymbol (SProxy :: SProxy "literal"))
```
Comment (rvion) [inline examples/passing/SolvingIsSymbol.purs]: How about checking the result too ?
```
let 
    a = (reflectSymbol (SProxy :: SProxy "literal"))
    b = (reflectSymbol (SProxy :: SProxy (TypeConcat "Hello" "World")))
when (a == "literal" && b == "HelloWorld") (log "Done")
```

Comment (LiamGoodacre) [inline examples/passing/SolvingIsSymbol.purs]: Sounds good, thanks

Comment (LiamGoodacre) [inline examples/passing/SolvingIsSymbol.purs]: (also I didn't realise outputting `Done` was interpretted as 'test passing')

File: src/Language/PureScript/TypeChecker/Entailment.hs [hunk 3]
```diff
@@ -103,6 +106,18 @@ data SolverOptions = SolverOptions
...
+-- | Build a type class instance dictionary for the `IsSymbol` class for a symbol
+isSymbolDictionary :: String -> TypeClassDict
+isSymbolDictionary sym =
+  TypeClassDictionaryInScope (Left dict) [] C.IsSymbol [TypeLevelString sym] Nothing
+  where
+    inst = ObjectLiteral
+      [ ("reflectSymbol", Abs (Left (Ident C.__unused)) (Literal (StringLiteral sym)))
+      ]
+    classTy = TypeConstructor (fmap coerceProperName C.IsSymbol)
+    expr = TypedValue False (Literal inst) (TypeApp classTy (TypeLevelString sym))
```
Comment (LiamGoodacre) [inline src/Language/PureScript/TypeChecker/Entailment.hs]: I wasn't sure about `TypedValue` - have I done this correctly?  I didn't work out what the boolean argument was for and I didn't see it mentioned in the doc-comment for the constructor.

**Comment by LiamGoodacre** [comment]
@paf31: making `TypeClassDictionaryInScope` store the `Expr` is going to require some reorganising of module dependencies.

The following is a trimmed down version of our modules and their dependencies:
```
TypeClassDictionaries

Environment
  TypeClassDictionaries

AST.Declarations
  TypeClassDictionaries
  Environment

AST
  AST.Declarations

Externs
  AST
  Environment
  TypeClassDictionaries

TypeChecker.Monad
  Environment
  TypeClassDictionaries

TypeChecker.Entailment
  AST
  Environment
  TypeChecker.Monad
  TypeClassDictionaries

TypeChecker.Types
  TypeChecker.Entailment

TypeChecker
  AST
  Environment
  TypeChecker.Monad
  TypeChecker.Types
  TypeClassDictionaries
```

We need `TypeClassDictionaries` to depend on `Expr` which is defined in `AST.Declarations`. This causes a cycle between `AST.Declarations`, `Environment`, and `TypeClassDictionaries`.

**Comment by LiamGoodacre** [comment]
An alternative might be to remove the name/expr from TypeClassDictionaryInScope and just pair the dict up with the Expr where needed.  I'll have a think about it.

**Comment by LiamGoodacre** [comment]
Although that is actually the same solution as what I've already done with adding the type parameter.

**Comment by paf31** [comment]
@LiamGoodacre I think we should stick with the original approach actually, using a sum type, now that I understand the problem properly (sorry!)

But let's rename some things to make them more understandable:

- `tcdName` should be named something like `tcdValue` since it's more general now.
- `NamedTypeClassDictionaryInScope` is a bit wordy. How about just `NamedDict` or something?

File: src/Language/PureScript/TypeClassDictionaries.hs [hunk 4]
```diff
@@ -21,4 +21,7 @@ data TypeClassDictionaryInScope
     -- | Type class dependencies which must be satisfied to construct this dictionary
     , tcdDependencies :: Maybe [Constraint]
     }
-    deriving (Show)
+    deriving (Show, Functor)
```
Comment (paf31) [inline src/Language/PureScript/TypeClassDictionaries.hs]: Might as well derive the usual `Functor`, `Foldable` and `Traversable` here.

**Comment by paf31** [comment]
A third option is a sum type like

```haskell
data Evidence
  = NamedDictionary Ident
  | IsSymbolDictionary String
```

right?

File: src/Language/PureScript/TypeChecker/Entailment.hs (hunk 3 above)
Comment (paf31) [inline src/Language/PureScript/TypeChecker/Entailment.hs]: Do we need `TypedValue` here? I think it can be removed actually.

**Comment by paf31** [comment]
I can confirm that this makes my [`generics-rep` PR](https://github.com/purescript/purescript-generics-rep/pull/5) build successfully.

I did run into one weird issue though. If I don't import `Data.Symbol` (and `IsSymbol` specifically) in my test suite, then I get the following error:

```
psc: An internal error ocurred during compilation: "Missing value in mnLookup"
```

File: src/Language/PureScript/TypeChecker/Entailment.hs (hunk 3 above)
Comment (LiamGoodacre) [inline src/Language/PureScript/TypeChecker/Entailment.hs]: It appeared to be required by `exprToCoreFn` in `src/Language/PureScript/CoreFn/Desugar.hs` here:
```haskell
exprToCoreFn ss com _  (A.TypeClassDictionaryConstructorApp name (A.TypedValue _ (A.Literal (A.ObjectLiteral vs)) _)) =
```

Otherwise it hits the error case:
```haskell
exprToCoreFn _ _ _ e =
    error $ "Unexpected value in exprToCoreFn mn: " ++ show e
```

Comment (paf31) [inline src/Language/PureScript/TypeChecker/Entailment.hs]: Maybe you could add a case for 

```
exprToCoreFn ss com _  (A.TypeClassDictionaryConstructorApp name (A.Literal (A.ObjectLiteral vs))) =
```

or use a pattern guard here then?

File: src/Language/PureScript/TypeClassDictionaries.hs (hunk 1 above)
Comment (paf31) [inline src/Language/PureScript/TypeClassDictionaries.hs]: Makes sense, thanks!

**Comment by paf31** [comment]
I think there is relatively little needed to push this over the finish line, right? Just

- Switching back to a sum type instead of using `Expr` directly.
- Dealing with the `TypedValue` issue.
- Maybe some renamings here and there (`tcdName` in particular)

I think once this is merged, we should release 0.10.3, since it will enable some nice library features which depend on `generics-rep` deriving (among others).

**Comment by LiamGoodacre** [comment]
Hey thanks for the review.  I've been away without internet for a few days so I haven't had a chance to work on this.  I should be able to get this done today/tomorrow though 👍

**Comment by LiamGoodacre** [comment]
I've been looking into the `Missing value in mnLookup` issue from the `renameModules` part of `moduleToJs`.  This occurs because the dictionaries are created inline and use the type class constructor (which hasn't necessarily been imported).  The easiest way to fix this is to not use the constructor and just have the object literal - is that okay to do?  This also means that I don't need to change `exprToCoreFn`.

**Comment by garyb** [comment]
I would say no, not really. The dictionaries are not going to be records when the typechecker gets fixed for #928.

**Comment by LiamGoodacre** [comment]
@garyb yeah that makes sense, thanks.  So we'll need to implicitly import the type class constructor during code gen - what would be the preferred way of doing that?

**Comment by garyb** [comment]
Hmm, that's a tricky one actually... I'm not too sure! Probably as part of the `CoreFn` desugaring process, if we can figure it out there? There used to be some extra import elaboration done in there, but it was removed after it got figured out propery during name desugaring instead. Pretty sure we won't be able to do it that early on in this case though, right?

File: src/Language/PureScript/CoreFn/Desugar.hs [hunk 5]
```diff
@@ -209,6 +212,7 @@ findQualModules decls =
   fqValues :: A.Expr -> [ModuleName]
   fqValues (A.Var q) = getQual' q
   fqValues (A.Constructor q) = getQual' q
+  fqValues (A.TypeClassDictionaryConstructorApp C.IsSymbol _) = getQual' C.IsSymbol
```
Comment (LiamGoodacre) [inline src/Language/PureScript/CoreFn/Desugar.hs]: This lets us implicitly add an import of `Data.Symbol` if the `IsSymbol` class is used.  The import also gets deduped later if it's already there.
Is this a good place to add this?

Comment (LiamGoodacre) [inline src/Language/PureScript/CoreFn/Desugar.hs]: Perhaps it's not great to explicitly mention `IsSymbol` here: it means we'd have to update multiple places if we're to do a similar instance solving in the future.  But it's possibly odd to do it for all constructor applications?

Comment (paf31) [inline src/Language/PureScript/CoreFn/Desugar.hs]: Could you please add a comment here explaining why this line is necessary?

File: src/Language/PureScript/TypeChecker/Entailment.hs [hunk 6]
```diff
@@ -37,11 +37,13 @@ import Language.PureScript.TypeClassDictionaries
 import Language.PureScript.Types
 import qualified Language.PureScript.Constants as C
 
+type TypeClassDict =
+  TypeClassDictionaryInScope (Either Expr (Qualified Ident))
```
Comment (paf31) [inline src/Language/PureScript/TypeChecker/Entailment.hs]: I think I'd rather use a sum type here, with a constructor for named instances, and one for symbol dictionaries. Then instead of using `either id` in `mkDictionary`, you would pattern match, and include the instance generation code there. `isSymbolDictionary` would just apply the appropriate constructor.

File: src/Language/PureScript/TypeChecker/Entailment.hs [hunk 7]
```diff
@@ -103,6 +105,15 @@ data SolverOptions = SolverOptions
   -- ^ Should the solver be allowed to defer errors by skipping constraints?
   }
 
+-- | Build a type class instance dictionary for the `IsSymbol` class for a symbol
+isSymbolDictionary :: String -> TypeClassDict
```
Comment (paf31) [inline src/Language/PureScript/TypeChecker/Entailment.hs]: I originally read this as a question: "is this a symbol dictionary?" 😄 

Maybe something like `makeIsSymbolDictionary` would be better?

**Comment by LiamGoodacre** [comment]
Currently updating the test example so that it depends on the implicit importing.

**Comment by LiamGoodacre** [comment]
Okay I'm confused: the test is passing on some build jobs and failing on others 😕.

**Comment by paf31** [comment]
I think you need to update the cabal file.

**Comment by paf31** [comment]
Nice work!
